apps/hash-graph/lib/authorization/src/backend.rs

```diff
@@ -0,0 +1,56 @@
+mod spicedb;
+
+use core::fmt;
+use std::error::Error;
+
+use error_stack::Report;
+
+pub use self::spicedb::{RelationshipFilter, SpiceDb, SpiceDbConfig};
+use crate::zanzibar::{
+    Affiliation, Consistency, GenericAffiliation, GenericResource, GenericSubject, Resource,
+    Subject, Tuple, Zookie,
+};
+
+/// A backend for interacting with an authorization system based on the Zanzibar model.
+pub trait AuthorizationBackend {
+    /// Returns if the [`Subject`] has the specified permission or relation to an [`Resource`].
+    async fn check<S, A, R>(
+        &self,
+        subject: &S,
+        affiliation: &A,
+        resource: &R,
+        consistency: Consistency<'_>,
+    ) -> Result<CheckResponse, Report<CheckError>>
+    where
+        S: Subject + Sync,
+        A: Affiliation<R> + Sync,
+        R: Resource + Sync;
+}
+
+/// Return value for [`AuthorizationBackend::check`].
+#[derive(Debug)]
+pub struct CheckResponse {
+    /// If the [`Subject`] has the specified permission or relation to an [`Resource`].
+    pub has_permission: bool,
+    /// A token to determine the time at which the check was performed.
+    pub checked_at: Zookie<'static>,
+}
+
+type StringTuple = Tuple<
+    GenericResource<String, String>,
+    GenericAffiliation<String>,
+    GenericSubject<GenericResource<String, String>, GenericAffiliation<String>>,
+>;
+
+#[derive(Debug)]
+pub struct CheckError {
+    pub tuple: StringTuple,
+}
```

Comment:
Errors are likely to be moved out to a different folder at a later point. I'm not 100% sure about the design of the errors but I wanted the tuple to be associated with the error itself.

Reply:
Good idea. I think maybe having the tuple as well as a reason enum (if there are multiple) would be appropriate in that case.

`AuthorizationBackend` could be generic over the `reason` enum, which must implement `Display`; that way different backends can have different errors.

Reply:
A generic `reason` would try to invent a new error-handling. For that, I'd use error-stack. It is, however, always known what tuple is checked.